rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_client_info.hpp

```diff
@@ -191,6 +116,11 @@ class ClientListener : public eprosima::fastdds::dds::DataReaderListener
     info_->response_subscriber_matched_count_.store(publishers_.size());
   }
 
+  size_t get_unread_responses()
+  {
+    return info_->response_reader_->get_unread_count();
```

Comment:
I don't  recommend this. This change could take into account same samples repeatly if they were not read.

Reply:
@richiware Please see [this discussion](https://github.com/ros2/rmw_connextdds/pull/76#discussion_r831933749) from which I understand the correct implementation is to return the total number of unread changes.

Reply:
It would be great to have some input from @alsora and/or @mauropasse on this, BTW

Reply:
@MiguelCompany @richiware our previous implementation relied simply on a counter to check unread events, instead of asking the DDS about unread messages. For example:
```
void on_data_available(eprosima::fastdds::dds::DataReader * reader)
{
  if (on_new_message_cb_) {
    on_new_message_cb_(user_data_, 1);
  } else {
    new_data_unread_count_++;
  }
}

// Then when setting the user callback:
void set_on_new_message_callback(const void * user_data, rmw_event_callback_t callback)
{
  if (new_data_unread_count_) {
    callback(user_data, new_data_unread_count_);
    new_data_unread_count_ = 0;
  }
}
```
The reason for having a different `new_data_unread_count_` than the unread samples on the DDS queue, is that this function (`set_on_new_message_callback`) is sometimes called very fast twice in a row. With our approach:
First time: `callback(user_data, new_data_unread_count_);`
Second time: the callback is not called, since `new_data_unread_count_ = 0`. 

With your approach here, calling `info_->response_reader_->get_unread_count();` could potentially (if `set_on_new_message_callback` is called many times fast) lead to problems. 
In our case, every call of the `callback(...)` pushes events into a queue, which are then processed (messages are read). With your approach here, the queue could potentially have dulicated/triplicated/.. events per unread sample, if we didn't have time to process them between calls.
I'd stay with the approach taken in this PR: https://github.com/ros2/rmw_fastrtps/pull/625

Reply:
@mauropasse The thing is that there is no guarantee that `on_data_available` is called once per sample.

Another thing to consider is the case where you have KEEP_LAST with a depth of 1. As explained on [this comment](https://github.com/ros2/rmw_connextdds/pull/76#discussion_r832469624), there is always a possibility that some samples that were notified to the callback will not actually be read by the application layer because of QoS settings (mainly KEEP_LAST history). The samples could be counted but then be pushed out of the cache by newer samples before they can be take()n out of it.

Reply:
What about using

```cpp
    return info_->response_reader_->get_unread_count(true);
```

? (i.e. [mark_as_read = true](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#_CPPv4NK8eprosima7fastdds3dds10DataReader16get_unread_countEb))

and keeping a count here?
does marking samples as read affect how `take()` work?

Reply:
@richiware @MiguelCompany was this issue resolved?

Reply:
Yes, it was. We did exactly what you mention, see [here](https://github.com/ros2/rmw_fastrtps/blob/1dcfe806874956158e1d93a77034c740e8790e19/rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_client_info.hpp#L120)